File: molecular/src/main/java/com/hartwig/actin/molecular/orange/evidence/actionable/VirusEvidence.java
package com.hartwig.actin.molecular.orange.evidence.actionable;

import java.util.List;

import com.google.common.collect.Lists;
import com.hartwig.actin.molecular.orange.datamodel.virus.VirusInterpreterEntry;
import com.hartwig.serve.datamodel.ActionableEvent;
import com.hartwig.serve.datamodel.ActionableEvents;

import org.jetbrains.annotations.NotNull;

public class VirusEvidence {

    @NotNull
    public static VirusEvidence create(@NotNull ActionableEvents actionableEvents) {
        return new VirusEvidence();
    }

    @NotNull
    public List<ActionableEvent> findMatches(@NotNull VirusInterpreterEntry virus) {
        List<ActionableEvent> applicableEvents = Lists.newArrayList();

        return applicableEvents;
    }
}

File: molecular/src/main/kotlin/com/hartwig/actin/molecular/evidence/actionability/VirusEvidence.kt
package com.hartwig.actin.molecular.evidence.actionability

import com.hartwig.actin.molecular.datamodel.wgs.driver.Virus
import com.hartwig.actin.molecular.datamodel.wgs.driver.VirusType
import com.hartwig.serve.datamodel.ActionableEvent
import com.hartwig.serve.datamodel.ActionableEvents
import com.hartwig.serve.datamodel.characteristic.TumorCharacteristicType

internal class VirusEvidence private constructor(
    private val hpvCharacteristics: List<ActionableEvent>,
    private val ebvCharacteristics: List<ActionableEvent>
) : EvidenceMatcher<Virus> {

    override fun findMatches(event: Virus): List<ActionableEvent> {
        return if (!event.isReportable) {
            emptyList()
        } else when (event.type) {
            VirusType.HUMAN_PAPILLOMA_VIRUS -> {
                hpvCharacteristics
            }

            VirusType.EPSTEIN_BARR_VIRUS -> {
                ebvCharacteristics
            }

            else -> {
                emptyList()
            }
        }
    }

    companion object {
        fun create(actionableEvents: ActionableEvents): VirusEvidence {
            val (hpvCharacteristics, ebvCharacteristics) = actionableEvents.characteristics()
                .fold(Pair(emptyList<ActionableEvent>(), emptyList<ActionableEvent>())) { acc, actionableCharacteristic ->
                    when (actionableCharacteristic.type()) {
                        TumorCharacteristicType.HPV_POSITIVE -> {
                            Pair(acc.first + actionableCharacteristic, acc.second)
                        }

                        TumorCharacteristicType.EBV_POSITIVE -> {
                            Pair(acc.first, acc.second + actionableCharacteristic)
                        }

                        else -> {
                            acc
                        }
                    }
                }
            return VirusEvidence(hpvCharacteristics, ebvCharacteristics)
        }
    }
}
